File: backend/app/domain/strategy_controls.py
from __future__ import annotations

from datetime import datetime, time
from enum import StrEnum
from uuid import UUID

from pydantic import Field, field_validator, model_validator

from ._base import DomainSchema, utc_now


class SessionName(StrEnum):
    PREMARKET = "premarket"
    REGULAR = "regular"
    AFTER_HOURS = "after_hours"


class Weekday(StrEnum):
    MON = "MON"
    TUE = "TUE"
    WED = "WED"
    THU = "THU"
    FRI = "FRI"


_WEEKDAY_ORDER: dict[Weekday, int] = {
    Weekday.MON: 0,
    Weekday.TUE: 1,
    Weekday.WED: 2,
    Weekday.THU: 3,
    Weekday.FRI: 4,
}


class SessionWindow(DomainSchema):
    session: SessionName = SessionName.REGULAR
    start: time
    end: time

    @model_validator(mode="after")
    def validate_window(self) -> "SessionWindow":
        if self.start >= self.end:
            raise ValueError("session window start must be before end")
        return self


# TradingHorizon lives here for historical reasons; the actual operator-facing
# horizon field lives on Deployment (backend/app/deployments/models.py).
# Deployment imports TradingHorizon from this module.
class TradingHorizon(StrEnum):
    SCALPING = "scalping"
    INTRADAY = "intraday"
    SWING = "swing"
    POSITION = "position"
    OTHER = "other"


class AllowedDirections(StrEnum):
    LONG = "long"
    SHORT = "short"
    BOTH = "both"


class SessionPreference(StrEnum):
    REGULAR_ONLY = "regular_only"
    REGULAR_AND_EXTENDED = "regular_and_extended"


class StrategyControlsVersion(DomainSchema):
    id: UUID
    strategy_controls_id: UUID
    version: int = Field(ge=1)
    name: str
    timeframe: str
    allowed_directions: AllowedDirections = AllowedDirections.LONG
    higher_timeframe_confirmation_required: bool = False
    session_preference: SessionPreference = SessionPreference.REGULAR_ONLY
    session_windows: list[SessionWindow] = Field(default_factory=list)
    avoid_first_minutes: int | None = Field(default=None, ge=0)
    no_new_entries_after: time | None = None
    force_flat_by: time | None = None
    time_based_exit_after_bars: int | None = Field(default=None, ge=1)
    time_based_exit_after_minutes: int | None = Field(default=None, ge=1)
    time_based_exit_after_days: int | None = Field(default=None, ge=1)
    cooldown_bars: int | None = Field(default=None, ge=0)
    cooldown_minutes: int | None = Field(default=None, ge=0)
    max_trades_per_session: int | None = Field(default=None, ge=1)
    max_trades_per_day: int | None = Field(default=None, ge=1)
    earnings_news_blackout_enabled: bool = False
    max_consecutive_losses_halt: int | None = Field(default=None, ge=1)
    skip_power_hour: bool = False
    day_of_week_restrictions: tuple[Weekday, ...] = Field(default_factory=tuple)
    feature_refs: list[str] = Field(default_factory=list)
    regime_filter_refs: list[str] = Field(default_factory=list)
    created_at: datetime = Field(default_factory=utc_now)

    @field_validator("day_of_week_restrictions", mode="before")
    @classmethod
    def normalize_day_restrictions(cls, v: object) -> tuple[Weekday, ...]:
        if v is None:
            return ()
        if isinstance(v, (list, tuple)):
            unique = list(dict.fromkeys(Weekday(d) for d in v))
            return tuple(sorted(unique, key=lambda d: _WEEKDAY_ORDER[d]))
        return v  # type: ignore[return-value]

    @model_validator(mode="after")
    def validate_cooldown(self) -> "StrategyControlsVersion":
        if self.cooldown_bars is not None and self.cooldown_minutes is not None:
            raise ValueError("use cooldown_bars or cooldown_minutes, not both")
        return self

    @model_validator(mode="after")
    def validate_time_based_exit(self) -> "StrategyControlsVersion":
        set_units = sum(
            1
            for value in (
                self.time_based_exit_after_bars,
                self.time_based_exit_after_minutes,
                self.time_based_exit_after_days,
            )
            if value is not None
        )
        if set_units > 1:
            raise ValueError(
                "time-based exit accepts at most one of bars / minutes / days"
            )
        return self

    @model_validator(mode="after")
    def validate_force_flat_after_no_new_entries(self) -> "StrategyControlsVersion":
        if (
            self.no_new_entries_after is not None
            and self.force_flat_by is not None
            and self.force_flat_by < self.no_new_entries_after
        ):
            raise ValueError(
                "force_flat_by must be at or after no_new_entries_after"
            )
        return self

File: frontend/src/api/schemas/strategyControls.ts
import { z } from "zod";

// ------------------------------------------------------------------
// Enums
// ------------------------------------------------------------------

export const WeekdaySchema = z.enum(["MON", "TUE", "WED", "THU", "FRI"]);
export type Weekday = z.infer<typeof WeekdaySchema>;

export const TradingHorizonSchema = z.enum([
  "scalping",
  "intraday",
  "swing",
  "position",
  "other",
]);
export type TradingHorizon = z.infer<typeof TradingHorizonSchema>;

export const AllowedDirectionsSchema = z.enum(["long", "short", "both"]);
export type AllowedDirections = z.infer<typeof AllowedDirectionsSchema>;

export const SessionPreferenceSchema = z.enum([
  "regular_only",
  "regular_and_extended",
]);
export type SessionPreference = z.infer<typeof SessionPreferenceSchema>;

export const SessionNameSchema = z.enum([
  "premarket",
  "regular",
  "after_hours",
]);
export type SessionName = z.infer<typeof SessionNameSchema>;

// ------------------------------------------------------------------
// SessionWindow
// ------------------------------------------------------------------

export const SessionWindowSchema = z.object({
  session: SessionNameSchema,
  start: z.string(),
  end: z.string(),
});
export type SessionWindow = z.infer<typeof SessionWindowSchema>;

// ------------------------------------------------------------------
// Draft (write model)
// ------------------------------------------------------------------

export const StrategyControlsDraftSchema = z.object({
  name: z.string().min(1).max(120),
  timeframe: z.string().min(1),
  allowed_directions: AllowedDirectionsSchema.default("long"),
  higher_timeframe_confirmation_required: z.boolean().default(false),
  session_preference: SessionPreferenceSchema.default("regular_only"),
  session_windows: z.array(SessionWindowSchema).default([]),
  avoid_first_minutes: z.number().int().min(0).nullable().optional(),
  no_new_entries_after: z.string().nullable().optional(),
  force_flat_by: z.string().nullable().optional(),
  time_based_exit_after_bars: z.number().int().min(1).nullable().optional(),
  time_based_exit_after_minutes: z.number().int().min(1).nullable().optional(),
  time_based_exit_after_days: z.number().int().min(1).nullable().optional(),
  cooldown_bars: z.number().int().min(0).nullable().optional(),
  cooldown_minutes: z.number().int().min(0).nullable().optional(),
  max_trades_per_session: z.number().int().min(1).nullable().optional(),
  max_trades_per_day: z.number().int().min(1).nullable().optional(),
  earnings_news_blackout_enabled: z.boolean().default(false),
  max_consecutive_losses_halt: z.number().int().min(1).nullable().optional(),
  skip_power_hour: z.boolean().default(false),
  day_of_week_restrictions: z.array(WeekdaySchema).default([]),
  feature_refs: z.array(z.string()).default([]),
  regime_filter_refs: z.array(z.string()).default([]),
});
export type StrategyControlsDraft = z.infer<typeof StrategyControlsDraftSchema>;

// ------------------------------------------------------------------
// Domain version (payload inside records)
// ------------------------------------------------------------------

export const StrategyControlsVersionSchema = z
  .object({
    id: z.string(),
    strategy_controls_id: z.string(),
    version: z.number().int(),
    name: z.string(),
    timeframe: z.string(),
    allowed_directions: AllowedDirectionsSchema,
    higher_timeframe_confirmation_required: z.boolean(),
    session_preference: SessionPreferenceSchema,
    session_windows: z.array(SessionWindowSchema).default([]),
    avoid_first_minutes: z.number().nullable().optional(),
    no_new_entries_after: z.string().nullable().optional(),
    force_flat_by: z.string().nullable().optional(),
    time_based_exit_after_bars: z.number().nullable().optional(),
    time_based_exit_after_minutes: z.number().nullable().optional(),
    time_based_exit_after_days: z.number().nullable().optional(),
    cooldown_bars: z.number().nullable().optional(),
    cooldown_minutes: z.number().nullable().optional(),
    max_trades_per_session: z.number().nullable().optional(),
    max_trades_per_day: z.number().nullable().optional(),
    earnings_news_blackout_enabled: z.boolean(),
    max_consecutive_losses_halt: z.number().int().nullable().optional(),
    skip_power_hour: z.boolean().default(false),
    day_of_week_restrictions: z.array(WeekdaySchema).default([]),
    feature_refs: z.array(z.string()).default([]),
    regime_filter_refs: z.array(z.string()).default([]),
    created_at: z.string(),
  })
  .passthrough();
export type StrategyControlsVersion = z.infer<typeof StrategyControlsVersionSchema>;

// ------------------------------------------------------------------
// Persistence record
// ------------------------------------------------------------------

export const StrategyControlsVersionRecordSchema = z
  .object({
    payload: StrategyControlsVersionSchema,
    saved_at: z.string(),
  })
  .passthrough();
export type StrategyControlsVersionRecord = z.infer<
  typeof StrategyControlsVersionRecordSchema
>;

// ------------------------------------------------------------------
// Library summary (list row)
// ------------------------------------------------------------------

export const StrategyControlsLibrarySummarySchema = z
  .object({
    strategy_controls_id: z.string(),
    name: z.string(),
    head_version_id: z.string().optional(),
    head_version_number: z.number().int(),
    is_default: z.boolean(),
    retired_at: z.string().nullable(),
    usage_count: z.number().int(),
  })
  .passthrough();
export type StrategyControlsLibrarySummary = z.infer<
  typeof StrategyControlsLibrarySummarySchema
>;

export const StrategyControlsLibraryListResponseSchema = z
  .object({
    libraries: z.array(StrategyControlsLibrarySummarySchema).default([]),
  })
  .passthrough();
export type StrategyControlsLibraryListResponse = z.infer<
  typeof StrategyControlsLibraryListResponseSchema
>;

// ------------------------------------------------------------------
// Version summary (history entry)
// ------------------------------------------------------------------

export const StrategyControlsVersionSummarySchema = z
  .object({
    version_id: z.string(),
    version: z.number().int(),
    saved_at: z.string(),
  })
  .passthrough();
export type StrategyControlsVersionSummary = z.infer<
  typeof StrategyControlsVersionSummarySchema
>;

// ------------------------------------------------------------------
// Full library detail
// ------------------------------------------------------------------

export const StrategyControlsLibrarySchema = z
  .object({
    strategy_controls_id: z.string(),
    name: z.string(),
    is_default: z.boolean(),
    retired_at: z.string().nullable(),
    head: StrategyControlsVersionRecordSchema,
    history: z.array(StrategyControlsVersionSummarySchema).default([]),
  })
  .passthrough();
export type StrategyControlsLibrary = z.infer<typeof StrategyControlsLibrarySchema>;

// ------------------------------------------------------------------
// used-by response
// ------------------------------------------------------------------

export const StrategyControlsUsedByResponseSchema = z
  .object({
    deployment_ids: z.array(z.string()).default([]),
  })
  .passthrough();
export type StrategyControlsUsedByResponse = z.infer<
  typeof StrategyControlsUsedByResponseSchema
>;
